core/trino-main/src/main/java/io/trino/operator/window/matcher/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,317 @@
+        }
+
+        @Override
+        protected Void visitIrEmptyPattern(IrEmptyPattern node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        protected Void visitIrAnchorPattern(IrAnchorPattern node, Void context)
+        {
+            switch (node.getType()) {
+                case PARTITION_START:
+                    instructions.add(new MatchStart());
+                    return null;
+                case PARTITION_END:
+                    instructions.add(new MatchEnd());
+                    return null;
+                default:
+                    throw new IllegalStateException("unexpected anchor type: " + node.getType());
+            }
+        }
+
+        @Override
+        protected Void visitIrExcludedPattern(IrExcludedPattern node, Void context)
+        {
+            instructions.add(new Save());
+            process(node.getPattern());
+            instructions.add(new Save());
+            return null;
+        }
+
+        @Override
+        protected Void visitIrPatternAlternation(IrPatternAlternation node, Void context)
+        {
+            checkArgument(node.getPatterns().size() > 1, "invalid pattern: alternation with single element");
+
+            List<IrRowPattern> parts = node.getPatterns();
+            List<Integer> jumpPositions = new ArrayList<>();
+
+            for (int i = 0; i < parts.size() - 1; i++) {
+                int splitPosition = instructions.size();
+                instructions.add(new SplitDummy());
+                Split.Builder builder = new Split.Builder();
+                builder.setFirst(instructions.size());
+                process(parts.get(i));
+                jumpPositions.add(instructions.size());
+                instructions.add(new JumpDummy());
+                builder.setSecond(instructions.size());
+                instructions.set(splitPosition, builder.build());
+            }
+
+            process(parts.get(parts.size() - 1));
+
+            for (int position : jumpPositions) {
+                instructions.set(position, new Jump(instructions.size()));
+            }
+
+            return null;
+        }
+
+        @Override
+        protected Void visitIrPatternConcatenation(IrPatternConcatenation node, Void context)
+        {
+            checkArgument(node.getPatterns().size() > 1, "invalid pattern: concatenation with single element");
+            return concatenation(node.getPatterns());
+        }
+
+        @Override
+        protected Void visitIrPatternPermutation(IrPatternPermutation node, Void context)
+        {
+            checkArgument(node.getPatterns().size() > 1, "invalid pattern: permutation with single element. run IrRowPatternFlattener first");
+
+            Iterator<List<Integer>> permutations = orderedPermutations(IntStream.range(0, node.getPatterns().size())
+                    .boxed()
+                    .collect(toImmutableList()))
+                    .iterator();
+
```

Comment:
This is confusing. The methods in the chain are not all at the same level. For instance, `iterator()` is a call on the result of `orderedPermutations`, but `boxed()` and `collect()` are part of the chain on `IntStream.range`

I would rewrite this method as follows:

```java
            List<Integer> indexes = IntStream.range(0, node.getPatterns().size())
                    .boxed()
                    .collect(toImmutableList());

            List<List<IrRowPattern>> permutations = orderedPermutations(indexes).stream()
                    .map(permutation -> permutation.stream()
                            .map(node.getPatterns()::get)
                            .collect(toImmutableList()))
                    .collect(toImmutableList());

            return alternation(permutations);
```